README.md
<h1 align="center"><img src="misc/AppIcon/AppIcon0128.png" width="128" height="128" alt="MyHappyList" title="MyHappyList"></h1>

[![](https://img.shields.io/github/release/ReFreezed/MyHappyList.svg)](https://github.com/ReFreezed/MyHappyList/releases/latest)

**<big>MyHappyList</big>** - a very simple [AniDB](https://anidb.net/) *MyList* manager for Windows.

[Forum thread](https://anidb.net/thread83307)

- [Features](#features)
- [Download](#download)
- [How to Update](#how-to-update)
- [Translations](#translations)
- [Issues](#issues)



## Features
- Add files to *MyList*.
- Mark files as watched.
- Calculate [ed2k hashes](https://wiki.anidb.info/w/Ed2k-hash).

And that's about it! If you're looking for something different,
have a look at [alternative software](https://anidb.net/perl-bin/animedb.pl?show=software).



## Download
See the [latest release](https://github.com/ReFreezed/MyHappyList/releases/latest) page.
No installation required - just unzip anywhere and run **MyHappyList.exe** !



## How to Update
Either update the program through `File` > `Update Program`,
or manually download the latest version and simply replace the old program folder with the new one.

### Updating from 1.0.x to 1.1 or Later
Move the `cache` and `local` folders from the old program folder to the new one to preserve settings etc.
These folders will automatically be moved to your
[`%APPDATA%`](https://en.wikipedia.org/wiki/Special_folder#File_system_directories)
folder next time you run MyHappyList, so you only ever have to do this once.



## Translations
MyHappyList currently supports English and Swedish.

If you want to help translating the program into your language you can make a copy of
[languages/en-US.txt](languages/en-US.txt) and edit the lines.
When you're done you can [submit a pull request](https://github.com/ReFreezed/MyHappyList/compare) or
send me the file in a [PM on AniDB](https://anidb.net/perl-bin/animedb.pl?show=msg&do.new=1&msg.to=ReFreezed)
and the translation will be in the next release.



## Issues

### Program does not start and there's no error message
There may be issues if there are special characters in the path to the program.
Try placing the program folder in `C:\Program Files`.

### Other issue
You can report the issue on [GitHub](https://github.com/ReFreezed/MyHappyList/issues)
or in the [forum thread](https://anidb.net/thread83307).




Changelog.txt
v1.2.5 (2019-12-30)
- Added option to mark a missing file as 'deleted' in MyList.
- Fixed an issue that sometimes made infinite error messages pop up when closing the program.

v1.2.4 (2019-11-09)
- Fixed an issue with paths containing non-ASCII characters.

v1.2.3 (2019-08-28)
- Hopefully fixed a network connection error on startup.
- Hopefully fixed the program sometimes silently continuing to run after it was supposed to close.

v1.2.2 (2019-06-30)
- Fixed localization related crash.
- Improved a network error message on startup.

v1.2.1 (2019-02-02)
- Fixed error on startup after an update.

v1.2.0 (2019-01-30)
- Added support for translations.
- Added setting: Server response timeout. Try increasing this if you keep getting timeout errors.
- Added button for restoring default settings.
- Files can now have the "remote storage" state in MyList, just like on the website.
- View date can now be specified for files.
- Fixed issue with the updater not being able to start the program after an update.

v1.1.1 (2018-08-30)
- Fixed updater not working properly.
- Fixed startup error if the 'Open' button was used in Windows Explorer.

v1.1.0 (2018-08-21)
- Added simple updating system.
- Saved files are now stored in %APPDATA% instead of the program folder.
- Added menu item for updating the status of selected files.
- Fixed a memory leak.

v1.0.4 (2018-08-14)
- Files that don't exist on AniDB are handled better.

v1.0.3 (2018-08-12)
- Fixed large files (4+ GB) not hashing correctly.
- Fixed crash when choosing the new location for moved files... again.
- Fixed keyboard focus being stolen when hashing.

v1.0.2 (2018-08-12)
- Fixed crash when choosing the new location for moved files.

v1.0.1 (2018-08-11)
- Super tiny general update.

v1.0.0 (2018-08-09)
- Initial release.


src/globals.lua
--[[============================================================
--=
--=  Global Values
--=
--=-------------------------------------------------------------
--=
--=  MyHappyList - manage your AniDB MyList
--=  - Written by Marcus 'ReFreezed' Thunström
--=  - MIT License (See main.lua)
--=
--============================================================]]



-- Settings, debug.

DEBUG                             = false
DEBUG_LOCAL                       = false and DEBUG

DEBUG_FORCE_NAT_OFF               = false and DEBUG_LOCAL
DEBUG_EXPIRATION_TIME_PORT        = 30
DEBUG_EXPIRATION_TIME_SESSION     = 3*60 -- Only useful is NAT is off.

DEBUG_DISABLE_VARIOUS_FILE_SAVING = false and DEBUG_LOCAL
DEBUG_NEVER_UP_TO_DATE            = false and DEBUG

-- Allow overriding DEBUG settings through a local file.
pcall(require, "local.debug")



-- Settings, app/general.

APP_NAME          = "MyHappyList"
APP_VERSION       = require"version"
MAX_DROPPED_FILES = 1000
SAVE_DELAY        = DEBUG and 1000 or 2000 -- Affects saving of settings, etc.

MARGIN_L          = 20
MARGIN_M          = 8
MARGIN_S          = 4
MARGIN_XS         = 2



-- Settings, AniDB.

SERVER_ADDRESS                      = DEBUG_LOCAL and "localhost" or "api.anidb.net"
SERVER_PORT                         = 9000
LOCAL_PORT                          = DEBUG_LOCAL and 9000 or 24040
LOCAL_PORT_FALLBACK                 = DEBUG_LOCAL and 9001 or 25050

PROTOCOL_VERSION                    = 3
CLIENT_NAME                         = "myhappylist"
CLIENT_VERSION                      = 2

DEFAULT_PING_DELAY                  = DEBUG_LOCAL and 20 or 60
DELAY_BETWEEN_REQUESTS_LONG         = 4
DELAY_BETWEEN_REQUESTS_SHORT        = 2
DEFAULT_SERVER_RESPONSE_TIMEOUT     = 120  -- Firefox's default HTTP response timeout used to be 120 seconds at some point.
DELAY_BEFORE_RESENDING              = {30, 2*60, 5*60, 10*60, 30*60, 1*60*60, 2*60*60, last=3*60*60}

NAT_LIMIT_MIN                       = 10
NAT_LIMIT_MAX                       = 15*60
NAT_LIMIT_TOLERANCE_BEFORE_SETTLING = 15

MAX_DATA_LENGTH                     = 1400 -- Must be between 400 and 1400. (MTU, PPPoE etc.)

FLOOD_PROTECTION_SHORT_TERM         = 20
FLOOD_PROTECTION_WINDOW             = 10   -- Saved amount of lastResponseTimes.



-- Constants.
require(... .."_keys") -- Extra stuff that WX don't add.
require(... .."_wx")

EMPTY_TABLE             = {}
NOOP                    = function()end

PROCESS_METHOD_ASYNC    = 1
PROCESS_METHOD_SYNC     = 2
PROCESS_METHOD_DETACHED = 3

--[[
	Directories.

	C:\ProgramData                   <- GetConfigDir
	C:\Users\User\AppData\Local      <- GetUserLocalDataDir
	C:\Users\User\AppData\Local\Temp <- GetTempDir
	C:\Users\User\AppData\Roaming    <- GetUserConfigDir, GetUserDataDir
	C:\Users\User\Documents          <- GetDocumentsDir
	<MyHappyList>\bin                <- GetDataDir, GetLocalDataDir, GetPluginsDir, GetResourcesDir
	<MyHappyList>\bin\wx.dll         <- GetExecutablePath
]]
local dirs = wxStandardPaths.Get()

local function absDirTo(dirMethod, subdir)
	return dirs[dirMethod](dirs):gsub("\\", "/") .. "/" .. APP_NAME .. subdir
end

DIR_APP        = wxGetCwd():gsub("\\", "/") -- Hopefully the program isn't launched from another folder...

DIR_CACHE      = absDirTo("GetUserDataDir",   (DEBUG_LOCAL and "/cacheDebug" or "/cache"))
DIR_CONFIG     = absDirTo("GetUserConfigDir", "")
DIR_LOGS       = absDirTo("GetUserDataDir",   "/logs")
DIR_TEMP       = absDirTo("GetTempDir",       "")

DIR_CACHE_OLD  = DEBUG_LOCAL and "cacheDebug" or "cache"
DIR_CONFIG_OLD = "local"
DIR_LOGS_OLD   = "logs"
DIR_TEMP_OLD   = "temp"

WRITABLE_DIRS = {
	absDirTo("GetUserConfigDir", ""),
	absDirTo("GetTempDir",       ""),
}

-- AniDB.

MYLIST_STATE_UNKNOWN                  = 0
MYLIST_STATE_INTERNAL_STORAGE         = 1
MYLIST_STATE_EXTERNAL_STORAGE         = 2
MYLIST_STATE_DELETED                  = 3
MYLIST_STATE_REMOTE_STORAGE           = 4

MYLIST_FILESTATE_NORMAL_ORIGINAL      = 0   -- normal
MYLIST_FILESTATE_CORRUPTED_OR_BAD_CRC = 1   -- normal
MYLIST_FILESTATE_SELF_EDITED          = 2   -- normal
MYLIST_FILESTATE_SELF_RIPPED          = 10  -- generic
MYLIST_FILESTATE_ON_DVD               = 11  -- generic
MYLIST_FILESTATE_ON_VHS               = 12  -- generic
MYLIST_FILESTATE_ON_TV                = 13  -- generic
MYLIST_FILESTATE_IN_THEATERS          = 14  -- generic
MYLIST_FILESTATE_STREAMED             = 15  -- normal
MYLIST_FILESTATE_ON_BLURAY            = 16  -- generic
MYLIST_FILESTATE_OTHER                = 100 -- normal



-- Modules.

local ok, _socket = pcall(require, "socket") -- Why the pcall? 2019-11-09
if ok then
	socket = _socket
	require"socket.url"
	require"socket.http"
end

_print = print -- We define our own print().



-- Variables and objects.

bypassDirectoryProtection = false
clearTempDirOnExit        = true
settingsAreFrozen         = true

logFile                   = nil
logFilePath               = ""

processes                 = {}


